src/frontend/js/clusive-message-queue.js

```diff
@@ -0,0 +1,144 @@
+/* global cisl, clusive, fluid_3_0_0, gpii, DJANGO_STATIC_ROOT, DJANGO_CSRF_TOKEN */
+
+(function(fluid) {
+    'use strict';
+
+    fluid.defaults("clusive.messageQueue", {       
+        gradeNames: ["fluid.component"],    
+        members: {
+            queue: []
+        },
+        config: {
+            // Interval for trying to flush queue
+            flushInterval: 60000,
+            // Namespace for local storage key
+            localStorageKey: "clusive.messageQueue.queue"
+        },
+        events: {
+            queueShouldFlush: null,
+            queueFlushSuccess: null,
+            queueFlushFailure: null,
+            syncedToLocalStorage: null        
+        },
+        listeners: {
+            "onCreate.syncFromLocalStorage": {
+                func: "clusive.messageQueue.syncFromLocalStorage",
+                args: ["{that}"]                
+            },
+            "onCreate.setFlushInterval": {
+                funcName: "{that}.setFlushInterval"                
+            },
+            "queueShouldFlush.flushQueue": {
+                funcName: "{that}.flush"
+            },
+            "queueFlushSuccess.clearQueue": {
+                func: "{that}.clearQueue",                
+            }           
```

Comment:
If I'm understanding correctly, the sequence is:

1. Timer invokes queue flush
2. Current message list read from localstorage; network connection to server initiated
3. After server round trip finishes, clearQueue is invoked
4. Content of local storage is erased.

What happens if between steps 2 and 3 (during the time we're waiting for a response from the server, which could be a few seconds) new messages are added to the queue? Wouldn't they be lost?

Reply:
Good point - writing some code to address this right now and will write a test that simulates a non-instant round trip to make sure this is addressed and there's not a chance of message loss.